jumanji/environments/routing/multi_cvrp/generator.py

```diff
@@ -0,0 +1,144 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.routing.multi_cvrp.types import (
+    Node,
+    PenalityCoeff,
+    State,
+    StateVehicle,
+    TimeWindow,
+)
+from jumanji.environments.routing.multi_cvrp.utils import (
+    DEPOT_IDX,
+    create_action_mask,
+    generate_uniform_random_problem,
+    get_init_settings,
+)
+
+
+class Generator(abc.ABC):
+    """Base class for generators for the MACVPR environment."""
+
+    def __init__(self, num_customers: int, num_vehicles: int) -> None:
+        """Initialises a MultiCVRP generator, used to generate the problem.
+
+        Args:
+            num_customers: number of customer nodes in the environment.
+            num_vehicles: number of vehicles in the environment.
+        """
+        self._num_customers = num_customers
+        self._num_vehicles = num_vehicles
+        # Scenario are taken from the paper
+        # Note: The time window detail could not be found in the paper for
+        # the 20, 50 and 150 customer scenarios. We use the 150 customer scenario's
+        # time window of 20 for them.
+        self._time_window_length = 20
+
+        (
+            self._map_max,
+            self._max_capacity,
+            self._max_start_window,
+            self._early_coef_rand,
+            self._late_coef_rand,
+            self._customer_demand_max,
+        ) = get_init_settings(self.num_customers, self.num_vehicles)
+
+        self._max_end_window = self._max_start_window + self._time_window_length
+
+    @property
+    def num_customers(self) -> int:
+        return self._num_customers
+
+    @property
+    def num_vehicles(self) -> int:
+        return self._num_vehicles
+
+    @abc.abstractmethod
+    def __call__(self, key: chex.PRNGKey) -> State:
+        """Generates a `MultiCVRP` state.
+
+        Returns:
+            A `MultiCVRP` state.
+        """
+
+
+class UniformRandomGenerator(Generator):
+    """Randomly (uniformly) places the customers and vehicles on a fixed size map."""
+
+    def __init__(self, num_customers: int, num_vehicles: int) -> None:
+        """Initialises a MultiCVRP generator, used to generate the problem.
+
+        Args:
+            num_customers: number of customer nodes in the environment.
+            num_vehicles: number of vehicles in the environment.
+        """
+        super().__init__(num_customers, num_vehicles)
+
+    def __call__(self, key: chex.PRNGKey) -> State:
+        """Generates a `MACVPR` state.
+
+        Returns:
+            A `MACVPR` state.
```

Comment:
Can we please update all occurrences of macvrp to multi_cvrp?
```suggestion
        """Generates a `MultiCVRP` state.

        Returns:
            A `MultiCVRP` state.
```